File: apps/admin-dashboard/lib/api-client.ts
import { HttpError, type SerializedHttpError } from "../../../packages/errors/src/base/http-error";
import { ERROR_CODES } from "../../../packages/errors/src/types/error-codes";
import {
  API_ROUTES,
  HEADER_AUTHORIZATION,
  HEADER_CORRELATION_ID,
  HEADER_TENANT_ID,
  type ApiServiceKey,
} from "./constants";
import { createCorrelationId } from "./helpers";

export class ApiClientError extends HttpError {
  static fromResponse(status: number, body: unknown): ApiClientError {
    const serialized = body as SerializedHttpError | {
      message?: string;
      error?: { message?: string; code?: string; metadata?: Record<string, unknown> };
    };

    const message =
      serialized?.error?.message ??
      (serialized as SerializedHttpError)?.error?.message ??
      ("message" in (serialized ?? {}) ? (serialized as { message?: string }).message : undefined) ??
      "Request failed";

    const code =
      serialized?.error?.code ??
      (serialized as SerializedHttpError)?.error?.code ??
      ERROR_CODES.SYSTEM.INTERNAL_SERVER_ERROR;

    return new ApiClientError({
      message,
      code: code as (typeof ERROR_CODES)["SYSTEM"]["INTERNAL_SERVER_ERROR"],
      statusCode: status,
      metadata: serialized?.error?.metadata ?? {},
    });
  }
}

export type ApiClientContext = {
  accessToken?: string | null;
  tenantId?: string | null;
  correlationId?: string | null;
};

export type ApiClientOptions = Omit<RequestInit, "body"> & {
  service: ApiServiceKey;
  path?: string;
  params?: Record<string, string | number | boolean | undefined>;
  context?: ApiClientContext;
  jsonBody?: unknown;
};

type TokenProvider = () => string | null;
type RefreshHandler = () => Promise<string | null>;
type TenantProvider = () => string | null;
type OnUnauthorized = () => void;

let accessTokenProvider: TokenProvider | null = null;
let refreshHandler: RefreshHandler | null = null;
let tenantIdProvider: TenantProvider | null = null;
let onUnauthorized: OnUnauthorized | null = null;
let refreshPromise: Promise<string | null> | null = null;

export function configureApiClient(config: {
  getAccessToken: TokenProvider;
  refreshAccessToken: RefreshHandler;
  getTenantId: TenantProvider;
  onUnauthorized?: OnUnauthorized;
}) {
  accessTokenProvider = config.getAccessToken;
  refreshHandler = config.refreshAccessToken;
  tenantIdProvider = config.getTenantId;
  onUnauthorized = config.onUnauthorized ?? null;
}

async function refreshAccessToken(): Promise<string | null> {
  if (!refreshHandler) {
    return null;
  }
  if (!refreshPromise) {
    refreshPromise = refreshHandler().finally(() => {
      refreshPromise = null;
    });
  }
  return refreshPromise;
}

function buildServiceUrl(
  service: ApiServiceKey,
  path = "",
  params?: ApiClientOptions["params"],
): string {
  const base = API_ROUTES[service];
  const normalizedPath = path.startsWith("/") ? path : path ? `/${path}` : "";
  const url = new URL(`${base}${normalizedPath}`, window.location.origin);

  if (params) {
    for (const [key, value] of Object.entries(params)) {
      if (value !== undefined) {
        url.searchParams.set(key, String(value));
      }
    }
  }

  return url.toString();
}

async function buildHeaders(
  context: ApiClientContext,
  headers?: HeadersInit,
): Promise<Record<string, string>> {
  const correlationId = context.correlationId ?? createCorrelationId();
  const accessToken = context.accessToken ?? accessTokenProvider?.() ?? null;
  const tenantId = context.tenantId ?? tenantIdProvider?.() ?? null;

  return {
    "Content-Type": "application/json",
    Accept: "application/json",
    [HEADER_CORRELATION_ID]: correlationId,
    ...(tenantId ? { [HEADER_TENANT_ID]: tenantId } : {}),
    ...(accessToken ? { [HEADER_AUTHORIZATION]: `Bearer ${accessToken}` } : {}),
    ...(headers as Record<string, string> | undefined),
  };
}

export function createApiClient(getContext: () => ApiClientContext) {
  async function request<T>(options: ApiClientOptions, retry = true): Promise<T> {
    const { service, path = "", params, context, headers, jsonBody, ...init } = options;
    const session = { ...getContext(), ...context };

    const response = await fetch(buildServiceUrl(service, path, params), {
      ...init,
      body: jsonBody !== undefined ? JSON.stringify(jsonBody) : undefined,
      headers: await buildHeaders(session, headers),
    });

    const payload = await response.json().catch(() => null);

    if (response.status === 401 && retry && refreshHandler) {
      const newToken = await refreshAccessToken();
      if (newToken) {
        return request<T>({ ...options, context: { ...session, accessToken: newToken } }, false);
      }
      onUnauthorized?.();
    }

    if (!response.ok) {
      throw ApiClientError.fromResponse(response.status, payload);
    }

    if (payload && typeof payload === "object" && "data" in (payload as Record<string, unknown>)) {
      return (payload as { data: T }).data;
    }

    return payload as T;
  }

  return {
    get: <T>(
      service: ApiServiceKey,
      path?: string,
      options?: Omit<ApiClientOptions, "service" | "path" | "method">,
    ) => request<T>({ ...options, service, path, method: "GET" }),
    post: <T>(
      service: ApiServiceKey,
      path?: string,
      jsonBody?: unknown,
      options?: Omit<ApiClientOptions, "service" | "path" | "jsonBody" | "method">,
    ) => request<T>({ ...options, service, path, jsonBody, method: "POST" }),
    put: <T>(
      service: ApiServiceKey,
      path?: string,
      jsonBody?: unknown,
      options?: Omit<ApiClientOptions, "service" | "path" | "jsonBody" | "method">,
    ) => request<T>({ ...options, service, path, jsonBody, method: "PUT" }),
    patch: <T>(
      service: ApiServiceKey,
      path?: string,
      jsonBody?: unknown,
      options?: Omit<ApiClientOptions, "service" | "path" | "jsonBody" | "method">,
    ) => request<T>({ ...options, service, path, jsonBody, method: "PATCH" }),
    delete: <T>(
      service: ApiServiceKey,
      path?: string,
      options?: Omit<ApiClientOptions, "service" | "path" | "method">,
    ) => request<T>({ ...options, service, path, method: "DELETE" }),
  };
}

File: apps/mobile/lib/services/api_client.dart
import 'package:dio/dio.dart';
import 'package:ordella_mobile/core/config/app_config.dart';
import 'package:ordella_mobile/core/constants/api_constants.dart';
import 'package:ordella_mobile/core/errors/api_exception.dart';
import 'package:uuid/uuid.dart';

typedef ApiContextProvider = ({
  String? accessToken,
  String? tenantId,
  String? correlationId,
}) Function();

typedef TokenRefreshCallback = Future<bool> Function();

class ApiClient {
  ApiClient({
    required AppConfig config,
    required ApiContextProvider getContext,
    TokenRefreshCallback? onUnauthorized,
    Dio? dio,
  })  : _getContext = getContext,
        _onUnauthorized = onUnauthorized,
        _dio = dio ??
            Dio(
              BaseOptions(
                baseUrl: config.apiGatewayUrl,
                connectTimeout: const Duration(seconds: 15),
                receiveTimeout: const Duration(seconds: 30),
                headers: {'Content-Type': 'application/json'},
              ),
            ) {
    _dio.interceptors.add(
      InterceptorsWrapper(
        onRequest: (options, handler) {
          final ctx = _getContext();
          final correlationId = ctx.correlationId ?? const Uuid().v4();
          options.headers[ApiConstants.correlationIdHeader] = correlationId;
          if (ctx.tenantId != null && ctx.tenantId!.isNotEmpty) {
            options.headers[ApiConstants.tenantHeader] = ctx.tenantId;
          }
          if (ctx.accessToken != null && ctx.accessToken!.isNotEmpty) {
            options.headers[ApiConstants.authorizationHeader] =
                'Bearer ${ctx.accessToken}';
          }
          handler.next(options);
        },
        onError: (error, handler) async {
          final status = error.response?.statusCode;
          if (status == 401 && _onUnauthorized != null) {
            final refreshed = await _onUnauthorized!();
            if (refreshed) {
              try {
                final response = await _retry(error.requestOptions);
                handler.resolve(response);
                return;
              } catch (_) {}
            }
          }
          handler.next(error);
        },
      ),
    );
  }

  final ApiContextProvider _getContext;
  final TokenRefreshCallback? _onUnauthorized;
  final Dio _dio;

  String servicePath(String service) {
    final path = ApiConstants.gatewayPaths[service];
    if (path == null) {
      throw ApiException('Unknown service: $service');
    }
    return path;
  }

  Future<T> get<T>(
    String service,
    String path, {
    Map<String, dynamic>? queryParameters,
  }) {
    return _request<T>(
      method: 'GET',
      service: service,
      path: path,
      queryParameters: queryParameters,
    );
  }

  Future<T> post<T>(
    String service,
    String path, {
    dynamic body,
    Map<String, dynamic>? queryParameters,
  }) {
    return _request<T>(
      method: 'POST',
      service: service,
      path: path,
      body: body,
      queryParameters: queryParameters,
    );
  }

  Future<T> patch<T>(
    String service,
    String path, {
    dynamic body,
    Map<String, dynamic>? queryParameters,
  }) {
    return _request<T>(
      method: 'PATCH',
      service: service,
      path: path,
      body: body,
      queryParameters: queryParameters,
    );
  }

  Future<T> delete<T>(
    String service,
    String path, {
    Map<String, dynamic>? queryParameters,
  }) {
    return _request<T>(
      method: 'DELETE',
      service: service,
      path: path,
      queryParameters: queryParameters,
    );
  }

  Future<T> _request<T>({
    required String method,
    required String service,
    required String path,
    dynamic body,
    Map<String, dynamic>? queryParameters,
  }) async {
    final normalizedPath = path.startsWith('/') ? path : '/$path';
    final url = '${servicePath(service)}$normalizedPath';

    try {
      final response = await _dio.request<dynamic>(
        url,
        data: body,
        queryParameters: queryParameters,
        options: Options(method: method),
      );

      return _unwrap<T>(response.data);
    } on DioException catch (e) {
      final payload = e.response?.data;
      final message = _extractMessage(payload) ?? e.message ?? 'Request failed';
      throw ApiException(message, statusCode: e.response?.statusCode, payload: payload);
    }
  }

  Future<Response<dynamic>> _retry(RequestOptions options) {
    final ctx = _getContext();
    final headers = Map<String, dynamic>.from(options.headers);
    if (ctx.tenantId != null) {
      headers[ApiConstants.tenantHeader] = ctx.tenantId;
    }
    if (ctx.accessToken != null) {
      headers[ApiConstants.authorizationHeader] = 'Bearer ${ctx.accessToken}';
    }

    return _dio.request<dynamic>(
      options.path,
      data: options.data,
      queryParameters: options.queryParameters,
      options: Options(
        method: options.method,
        headers: headers,
        contentType: options.contentType,
        responseType: options.responseType,
      ),
    );
  }

  T _unwrap<T>(dynamic data) {
    if (data == null) {
      return null as T;
    }
    if (data is Map<String, dynamic> && data.containsKey('data')) {
      return data['data'] as T;
    }
    return data as T;
  }

  String? _extractMessage(dynamic payload) {
    if (payload is Map<String, dynamic>) {
      final message = payload['message'];
      if (message is String) return message;
      if (message is List && message.isNotEmpty) {
        return message.first.toString();
      }
    }
    return null;
  }
}
